src/sensitivities/chainrules.jl

```diff
@@ -0,0 +1,328 @@
+
+    # preprocess has a similar definition, signature-wise, to what is in signature_def
+    preprocess_def = Dict{Symbol, Any}(
+        :name => :preprocess,
+        :args => [op, :($y::Branch), ȳ, args...],
+        :body => quote
+            pullback = getfield($y, :pullback)  # avoid issues with getproperty overloading
+            @assert(pullback !== nothing, "pullback not set, probably because different code path used for preprocess vs for ∇. Probably need to delete a defination for ∇")
+            return pullback($ȳ)
+        end,
+    )
+
+    where_params = get(signature_def, :whereparams, nothing)
+    if where_params !== nothing
+        preprocess_def[:whereparams] = where_params
+    end
+    return ExprTools.combinedef(preprocess_def)
+end
+
+"""
+    ∇_declaration(unionised_signature_def)
+
+Generates that AST for the overload of the `∇` function which returns the gradient for
+specified arguments.
+Basically this generates things like:
+`∇(::opT, ::Type{Arg{N}}, p, y, ȳ, xs...) where N = p[N+1]  # Skip dself`
+where `p` is the pullback computed by [`preprocess`](@ref)
+
+Note that the `unionised_signature_def` must already have been unionised to accept `Node`s.
+"""
+function ∇_declaration(signature_def)
+    # For readability lets name all the parts, NB: this is being a bit too cute.
+    op = signature_def[:name]
+    args = signature_def[:args]
+
+    N = gensym(:N)
+    p = gensym(:p)
+    y = :(::Any)
+    ȳ = :(::Any)
+
+    ∇_def = Dict{Symbol, Any}(
+        :name => :∇,
+        :args => [op, :(::Type{Arg{$N}}), p, y, ȳ, args...],
+        :whereparams => [N; get(signature_def, :whereparams, [])],
+        :body => quote $p[$N+1] end,  # skip dself
+        :kwargs => [:(kwargs...)],
+    )
+    return ExprTools.combinedef(∇_def)
+end
+
+
+"""
+    _args_tuple(arg_exprs)
+
+For `arg_exprs` being a list of arguments expressions from a signature, of a form
+such as `[:(x::Int), :(y::Float64), :(z::Vararg)]`, returns a tuple expresion containing all
+of them by name; while correctly handling splatting, for things that are `Vararg` typed.
+e.g for the prior example `:((x, y, z...))`
+"""
+function _args_tuple(arg_exprs)
+    ret = Expr(:tuple)
+    ret.args = map(arg_exprs) do arg
+        @assert Meta.isexpr(arg, :(::), 2)
+        arg_name, Texpr = arg.args
+        if Meta.isexpr(Texpr, :where)  # remove where from `Vararg{T, N} where {T, N}`
+            Texpr = Texpr.args[1]
+        end
+        # Needs to be after removing `where`
+        if Meta.isexpr(Texpr, :curly) # remove `{T, N}` from `Vararg{T,N```
+            Texpr = Texpr.args[1]
+        end
+        if Texpr == :Vararg
+            return :($arg_name...)
+        else
+            return arg_name
+        end
+    end
+    return ret
+end
+
+"""
+    _truely_rename_unionall(@nospecialize(u))
+
+For `u` being a `UnionAll` this replaces every `TypeVar` with  a new one with a `gensym`ed
+names. This is useful for manual macro-hygine.
```

Comment:
manual macro-hygine.
Some code clashes because a type var is being generated that has same name as one that is writen into the code.
What should I add to this doc string to help make that clearer?
I thought the mention of `this is useful for manual macro-hygine` is enough.
Do I need to clarify why hygine is useful, or why it is useful in thsis particular case or somethign else?

Reply:
`gensym_unionall` maybe a better name. Could be a candidate for ExprTools.jl but I'm still unclear on why you need this

Reply:
It is useful because if the code is generating `foo(x::(Vector{T} where T), y::(Vector{T} where T), z::T) where T`.
There are 3 distince type variables `T` in that.
but julia is not OK about this fact. I think it is actually a bug in the language itself, but not one a human is likely to hit.
https://github.com/JuliaLang/julia/issues/39876
So if one renames the the T's that are in the 2 UnionAlls to be something like `#T#1` and `#T#2`, this fixes it.

Reply:
This has moved to ExprTools as the `extra_hygiene` flag.
Further discussion on it should happen there